Template Pipeline: Ingestion of bindings on `ng-template`s

dylhunn opened the pull request:
Consider code like the following:

```
<ng-template literal1="litVal" [prop1]="propVal" [attr.attr1]="attrVal" [class.class1]="classVal">
  <h1 *ngIf="true" literal2="litVal" [prop2]="propVal" [attr.attr2]="attrVal" [class.class2]="classVal">
  </h1>
</ng-template>
```
It's quite difficult to follow how the template pipeline ingests these bindings today. `attrs` vs `templateAttrs`, text vs const bindings, structural vs raw templates, bindings on templates that don't target them, property bindings in the `Template` const field, etc. It's overwhelming.

Fear no longer! This PR is dedicated to fixing ingestion of bindings on templates. The end goal is to make binding ingestion philosophically consistent and understandable when reading. Also, I hope to put a stop to the presubmit failures related to template bindings.

Several commits fix individual, specific issues with bindings on templates. Along the way, I have performed major refactors of the binding ingestion code. This occasionally touches ops and phases, although the bulk of the work is in `ingest.ts`.

Read individual commit messages in order, for details.

dylhunn committed:
refactor(compiler): Don't emit class and style bindings on structural template views

The Template Pipeline has had a number of tricky bugs involving bindings on structural elements.

Consider this template:

```
<div *ngIf="true" [class.bar]="field"></div>
```

We were incorrectly emitting `ɵɵclassProp` on *both* the template's view, and the inner view. The solution is to just emit an extracted attribute on the enclosing template, so it still shows up in the const array, but does not affect the update block.

We will refactor binding ingestion soon, but this commit improves our correctness before any big refactor.

dylhunn committed:
refactor(compiler): Simplify some code in ingest

Currently Template Pipeline's ingest phase is very complex, especially when it comes to ingesting bindings.

In this commit, we make some superficial simplifications, in preparation for a larger refactoring. For example, we pull out common code such as `convertAstWithInterpolation` and the `i18n.Message` checks. This enormously shrinks the main binding ingestion functions.

In addition, we reorder the binding kind and flags code above `ingestBindings`, so that `ingestBindings` and `ingestBinding` can be viewed together.

dylhunn committed:
refactor(compiler): Split up binding ingestion for elements and templates

Previously, we had `ingestBindings` and `ingestBinding`, which required tons of cases to support both elements and templates.

Now, we have two separate functions, `ingestElementBindings` and `ingestTemplateBindings`.

Thanks to the previous refactoring work, `ingestBinding` is now extremely compact. In fact, it's so compact that, in the elements case, it can just be inlined! Therefore, element binding ingestion is now quite easy to read.

The template case continues to be pretty gnarly, although I have already removed some code. In subsequent commits, we will simplify it even further.

dylhunn committed:
refactor(compiler): Eliminate `BindingFlags` and reorder parameters of `createTemplateBinding`

It turns out that `BindingFlags.BindingTargetsTemplate` is actally a redundant property! It will be true in either of the following cases:
1. The template is a normal non-structural `ng-template`. We already know this from `TemplateKind`.
2. The binding came from `templateAttrs` (instead of `attrs`). We have this information in `BindingFlags.IsStructuralTemplateAttribute`.

Therefore, I can just eliminate `BindingFlags.BindingTargetsTemplate`. There's no reason to keep `BindingFlags` around for a single value, so I convert `BindingFlags.IsStructuralTemplateAttribute` to a boolean parameter (with the eventual goal of eliminating it entirely).

Additionally, because element binding ingestion now calls `ir.createBindingOp` inline, it was difficult to compare it to template binding ingestion, which uses the  `createTemplateBinding` helper. I have changed the parameter order of `createTemplateBinding` to closely mimic `ir.createBindingOp`. This will both make the code easier to read, and allow me to easily replace one with the other in the future.

Lastly: the template binding ingestion function is the site of much of the binding ingestion complexity. Add an explanatory function comment.

dylhunn committed:
refactor(compiler): Fix some behavior with bindings on explicit `ng-template` elements

The behavior of explicit bindings on `ng-template`s was untested, and we differed from `TemplateDefinitionBuilder` significantly. We now have much more similar behavior, although not 100% identical.

For example, consider this templarte:
```
<ng-template l="l1" [p]="p1" [attr.a]="a1" [class.c]="c1"></ng-template>
```

It's not clear what a class binding on an `ng-template` would actually do. Nonetheless, it's well-defined behavior in TemplateDefinitionBuilder, which emits `property` instructions for all three bindings, and people actually do this in google3.

Note that some of these bindings don't really make much sense, but we have to support them for compatibility purposes.

See comments for an in-depth explanation of all the logic.

Also, add a test to exercise the problematic case.

dylhunn committed:
refactor(compiler): Fix animation bindings on structural elements

Consider the case:

```
<button *ngIf="true" [@anim]="field"></button>
```

Only the inner `button` should recieve a `property` instruction for the animation binding. We were previously emitting one for the implicit `ng-template` as well, and collecting it into the consts for the `ng-template`. Both of these issues are now fixed.

mmalerba commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -940,34 +917,69 @@ function ingestTemplateBindings(
+ *     string literal if the parser considered it a text binding.
+ * @param unit If the binding has a unit (e.g. `px` for style bindings), then this is the unit.
+ * @param securityContext The security context of the binding.
+ * @param isTextBinding Whether the binding is a text binding according to the parser (e.g. `<div
```
Could `isTextBinding` just be inferred based on whether value is a string literal? From the descriptions here it sounds like it could, or do we think its better to keep it explicit?

dylhunn replied:
Yes! Nice simplification

mmalerba commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -940,34 +917,69 @@ function ingestTemplateBindings(
+ * @param isStructuralTemplateAttribute Whether this binding actually applies to the structural
+ *     ng-template. For example, an `ngFor` would actually apply to the structural template. (Most
+ *     bindings on structural elements target the inner element, not the template.)
+ * @param templateKind Whether this is an explicit `ng-template` or an implicit template created by
```
or a block template

dylhunn replied:
This should actually never be a block template in this function. I'll add an explanatory comment.

mmalerba commented on packages/compiler/src/template/pipeline/src/ingest.ts:
```diff
@@ -899,15 +899,17 @@ function ingestTemplateBindings(
           output.sourceSpan));
     }
+    if (templateKind === ir.TemplateKind.Structural && output.type === e.ParsedEventType.Regular) {
+      // Animation bindings are excluded from the structural template's const array.
```
I feel like the comment and code are saying slightly different things here (though they may be functionally the same) maybe the code should read `&& output.type !== Animation`?

dylhunn replied:
Yep, I have flipped the condition to make them match.

mmalerba reviewed (approved):
LGTM with some minor comments. Awesome work, this is way easier to follow now!

dylhunn committed:
refactor(compiler): Make some fixups to template binidng ingestion

Responding to comments on #53457 about the previous sequence of template binding PRs.

dylhunn commented:
caretaker: g3 presubmit is green; recent changes are trivial and no need to run another one.

alxhub commented:
This PR was merged into the repository by commit 921964220dcbcb0e864d20d4c98bce4f100024db.
